micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java

```diff
@@ -143,21 +144,39 @@ public void bindTo(MeterRegistry registry) {
                             final Map<String, MemoryUsage> before = gcInfo.getMemoryUsageBeforeGc();
                             final Map<String, MemoryUsage> after = gcInfo.getMemoryUsageAfterGc();
 
-                            if (oldGenPoolName != null) {
-                                final long oldBefore = before.get(oldGenPoolName).getUsed();
-                                final long oldAfter = after.get(oldGenPoolName).getUsed();
-                                final long delta = oldAfter - oldBefore;
-                                if (delta > 0L) {
-                                    promotedBytes.increment(delta);
+                            if (!oldGenPoolNames.isEmpty()) {
+                                boolean isOldGenGarbageCollected = false;
+                                long liveDataSizeValue = 0L;
+                                long maxDataSizeValue = 0L;
+
+                                for (String oldGenPoolName : oldGenPoolNames) {
+                                    final long oldBefore = before.get(oldGenPoolName).getUsed();
+                                    final long oldAfter = after.get(oldGenPoolName).getUsed();
+                                    final long delta = oldAfter - oldBefore;
+                                    if (delta > 0L) {
+                                        promotedBytes.increment(delta);
+                                    }
+
+                                    // Some GC implementations such as G1 can reduce the old gen size as part of a
+                                    // minor GC. To track the live data size we record the value if we see a
+                                    // reduction in the old gen heap size or after a major GC.
+                                    if (oldAfter < oldBefore || GcGenerationAge.fromName(notificationInfo.getGcName()) == GcGenerationAge.OLD) {
+                                        isOldGenGarbageCollected = true;
+                                        final long oldMaxAfter = after.get(oldGenPoolName).getMax();
+
+                                        if (liveDataSizeValue > 0) {
```

Comment:
This condition will always be false because the variable is initialized with 0 and only set inside this conditional.